### PR TITLE
Add cf-style-provider

### PR DESCRIPTION
**Pull request by tajo**
Explained here https://github.com/cloudflare/cf-ui/issues/123.

It's already dogfooded by the `/example` and tests (felaTestContext). Once published, I will introduce it into cf-styleguide, cf-ux and cf-www-next as well.

### CODE REVIEW

File: packages/cf-style-provider/src/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  return createFelaRenderer({
+    plugins,
+    enhancers,
+    selectorPrefix: [usedOpts.selectorPrefix]
+  });
+};
+
+export const createProvider = opts => {
```
Comment (thejameskyle) [inline packages/cf-style-provider/src/index.js]: Why is this a HOC?

Comment (tajo) [inline packages/cf-style-provider/src/index.js]: Because you need to configure the [renderer](https://github.com/rofrischmann/fela/blob/master/docs/basics/Renderer.md). It doesn't seem like something that should be done in `render()` ?

File: packages/cf-style-provider/src/index.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  const renderer = createRenderer(opts);
+  const cssNode = opts && opts.cssNode ? opts.cssNode : undefined;
+  return class StyleProvider extends Component {
```
Comment (sejoker) [inline packages/cf-style-provider/src/index.js]: could use functional component instead

File: packages/cf-style-container/package.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
   },
   "dependencies": {
     "css-color-function": "^1.3.0",
+    "fela": "^4.3.2",
```
Comment (sejoker) [inline packages/cf-style-container/package.json]: what's the difference between `cf-style-container` and `cf-style-provider`. To me they could be merged into single component.

Comment (sejoker) [inline packages/cf-style-container/package.json]: please lock dependencies with yarn.lock. We don't want failed CI builds every time fela upgrades

File: packages/cf-style-provider/package.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (sejoker) [inline packages/cf-style-provider/package.json]: please lock dependencies with yarn.lock. We don't want failed CI builds every time fela upgrades

File: packages/cf-style-provider/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+```jsx
+import ReactDOMServer from 'react-dom/server';
+import { Provider } from 'react-fela';
+import { createRenderer } from 'cf-style-provider';
+
+// default values
+const renderer = createRenderer({
+  cssNode: null, // where to render CSS styles (DOM node)
+  fontNode: null, // where to render CSS styles related to fonts (DOM node)
+  dev: false, // in dev mode it applies more Fela plugins (validator, ...)
```
Comment (sejoker) [inline packages/cf-style-provider/README.md]: beautifier and monolithic as well. It would be useful explaining why we need them.

File: packages/cf-style-container/package.json (hunk 3 above)
Comment (tajo) [inline packages/cf-style-container/package.json]: `cf-style-container` is a set of helpers that are used **by Fela components**. `cf-style-provider` is used **by consumers of cf-ui**. We definitely don't want to merge them because `cf-style-provider` has a lot of Fela dependencies (plugins, enhancers etc) that have no use for components. They don't overlap at all.

File: packages/cf-style-provider/package.json (hunk 4 above)
Comment (tajo) [inline packages/cf-style-provider/package.json]: I would like to but we depend on lerna here. The yarn support is [experimental](https://github.com/lerna/lerna/issues/371), let's wait when it's more stable.

> We don't want failed CI builds every time fela upgrades

It should not be *every* time.

File: packages/cf-style-provider/README.md (hunk 5 above)
Comment (tajo) [inline packages/cf-style-provider/README.md]: done

File: packages/cf-style-provider/src/index.js (hunk 2 above)
Comment (tajo) [inline packages/cf-style-provider/src/index.js]: done

File: packages/cf-style-provider/package.json (hunk 4 above)
Comment (sejoker) [inline packages/cf-style-provider/package.json]: I could test [yarn support|https://github.com/lerna/lerna/releases/tag/v2.0.0-beta.38]. It allows install with yarn and nothing else, this is exactly what we need.

File: packages/cf-style-provider/package.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  "jsnext:main": "es/index.js",
+  "author": "Vojtech Miksu <vojtech@cloudflare.com>",
+  "license": "BSD-3-Clause",
+  "publishConfig": {
+    "registry": "http://registry.npmjs.org/"
+  },
+  "dependencies": {
+    "cf-style-const": "^0.3.0",
+    "cf-style-container": "^0.2.0",
+    "fela": "^4.3.2",
```
Comment (sejoker) [inline packages/cf-style-provider/package.json]: An idea: could we declare all fela dendencies as optional and keep them in the root package.json?

File: packages/cf-style-provider/package.json (hunk 4 above)
Comment (tajo) [inline packages/cf-style-provider/package.json]: Ok, but let's do in a different PR. There should be yarn.lock in every package. Update travis.yml. Test that it works for private packages as well (cf-ux).

Comment (sejoker) [inline packages/cf-style-provider/package.json]: @tajo agree.

File: packages/cf-style-provider/package.json (hunk 6 above)
Comment (tajo) [inline packages/cf-style-provider/package.json]: Root package.json? Do you mean cf-ui/package.json? I don't think so. 

Pretty much the whole point of cf-style-provider is to shield you from setting up Fela if you want to consume cf-ui components and provide you with all necessary dependencies.

File: packages/cf-style-provider/src/index.js (hunk 1 above)
Comment (thejameskyle) [inline packages/cf-style-provider/src/index.js]: I don't know why, they are both done statically. It's not going to run multiple times.

**Comment by tajo** [comment]
I removed HOC and simplified usage of `StyleProvider` as @thejameskyle suggested. @jwineman and @sejoker please take a quick look again.

**Comment by jwineman** [comment]
LGTM

**Comment by sejoker** [comment]
LGTM
